V2: Improve datasplit function for multicomponent dataset

hwpang opened the pull request:
## Description
Texts below are copied from https://github.com/chemprop/chemprop/pull/500#issuecomment-1771956143.

During working on an example notebook for multicomponent training, there are several things I noticed that the split_data is currently awkward and can be improved for multi-components.

With https://github.com/chemprop/chemprop/pull/501, a MulticomponentDataset dataset contains K datasets where K is the number of components. These K datasets need to be split consistently. The current split_data function is not able to do that. I discussed this with @KnathanM and we thought of some ways to do this, and I implemented one of them. Here are the possible options we came up with:

Option 1: Generalize split_data to be able to take in K lists of datapoints, use one of them to do the splitting, apply the same splitting to the rest of lists of datapoints. Supply split_key_molecule to indicate which one is used for splitting.

Option 2: Deal with this analogous to data scaler. Make split_data also return the split indices. If train_val_test_indices are provided, split the datapoints. This is the approach I took.

I also had additional discussion with @JacksonBurns today in person and we will work on this together to improve data split for multicomponent.

JacksonBurns commented:
Blocked by #510

JacksonBurns commented:
@hwpang the docstring for MulticomponentDataset is out of date

hwpang commented:
@JacksonBurns and I had a long in person discussion and we decided to only keep the changes that allow return as indices, and let the user take over from that as I did in the example notebook in https://github.com/chemprop/chemprop/pull/523.

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -60,6 +64,8 @@ def split_data(
         Unsupported split method requested
     """
+    if (num_splits := len(sizes)) != 3:
+        raise RuntimeError(f"Specify sizes for train, validation, and test (got {num_splits} values).")
```
```suggestion
        raise ValueError(f"Specify sizes for train, validation, and test (got {num_splits} values).")
```

davidegraff reviewed (commented):
Would it be possible to make `split_data()` always return indices and abstract out the actual splitting into higher-level functions? By that I mean something like the following functions that would get called from `train.py`:
```python
MulticomponentDatapoint = Sequence[MoleculeDatapoint]

def split_data(
    datapoints: Sequence[MoleculeDatapoint], *args, **kwargs
) -> tuple[tuple[int], tuple[int], tuple[int]]:
    # same as now but always returns indices


def split_monocomponent(
    datapoints: Sequence[MoleculeDatapoint], *args, **kwargs
) -> tuple[Sequence[MoleculeDatapoint], ...]:
    train_idxs, val_idxs, test_idxs = split_data(datapoints, *args, **kwargs)
    train = [datapoints[i] for i in train_idxs]
    # same for val and test

    return (train, val, test)


def split_multicomponent(
    datapoints: Sequence[MulticomponentDatapoint], key_index: int = 0, *args, **kwargs
) -> tuple[Sequence[MulticomponentDatapoint], ...]:
    key_datapoints = [datapoint[key_index] for datapoint in datapoints]
    train_idxs, val_idxs, test_idxs = split_data(key_datapoints, *args, **kwargs)
    train = [datapoints[i] for i in train_idxs]
    # same for val and test

    return (train, val, test)
```
These names are just suggestions--I don't actually care what they're called. But I think this design is slightly more idiomatic and clear what's going on because in `train.py` you're now able to split the data like so:
```python
if IS_MULTICOMPONENT:
    train, val, test = split_multicomponent(all_data)
else:
    train, val, test = split_monocomponent(all_data)
```
where `IS_MULTICOMPONENT` is some logical condition for detecting multicomponent input data

hwpang commented:
@davidegraff Thanks! I've modified the `split_data` to always return indices and abstract out the actual splitting as you suggested. Note that `split_data` can either return a tuple of 3 lists (training, validation, and testing), or it can return a tuple of 3 lists of sublists, where the sublists correspond to each fold of the cross validation.

I also change `split_multicomponent` because for multicomponent, `len(datapointss) == K` and `len(datapointss[0]) == N`, where `K` is the number of components and `N` is the size of dataset.

hwpang commented:
Rebased with v2/dev

hwpang commented:
@davidegraff Hey David, I want your opinion on updating the unit tests. Now that split_data returns indices but not data points, I need to update the unit tests. I can either keep it mostly as it is and just test that the indices are correct, or I can change the unit test to test the higher level function `split_monocomponent` and `split_multicomponent`? I think the former makes more sense to me as the higher level function just unpacks the indices to data points, but would like your thoughts on this.

davidegraff commented:
> @davidegraff Hey David, I want your opinion on updating the unit tests. Now that split_data returns indices but not data points, I need to update the unit tests. I can either keep it mostly as it is and just test that the indices are correct, or I can change the unit test to test the higher-level function `split_monocomponent` and `split_multicomponent`? I think the former makes more sense to me as the higher-level function just unpacks the indices to data points, but would like your thoughts on this.

I agree. The principled development model here would be to do this and "mock" the `split_data` data function in the client `split_{mono,multi}component` functions. That is, these functions are clients of the `split_data` data, so the logic they implement **must assume** that `split_data` works as intended, i.e., does what the docstring says it does. Of course, we write unit tests for this function to ensure that, but when those tests fail, it will cause unit tests of client functions to also fail. But are these unit tests failing because the client function logic is wrong or because the "server-side" code they're calling is wrong? It's hard to say, so these failures don't tell us much.

[Mocking](https://en.wikipedia.org/wiki/Mock_object) is a concept in unit tests that solves this issue. In this paradigm, we use `Mock` objects that "mock" the stated behavior of the object that they're mocking. Essentially, an object-oriented way to "hard-code" correct behavior. The benefit of hard-coding correct behavior is that for testing code that is a client of other code, the tests no longer depend on the server-side code being correct. A concrete example here would be to mock the behavior of `split_data` such that we can rely on it being correct for downstream testing of `split_{multi,mono}component`.

**However**, that's a good amount of work, so I don't have that expectation of dev's because we're not full-time SWEs. So all of this is to say, focus on unit-testing `split_data`, and at a later date, we can write unit tests for the client functions.

davidegraff commented on chemprop/v2/data/splitting.py:
```diff
@@ -124,31 +127,31 @@ def split_data(
             train = [
-                datapoints[ii]
+                ii
                 for ii in itertools.chain.from_iterable(
                     smiles_indices[unique_smiles[i]] for i in train_idxs
                 )
             ]
             val = [
-                datapoints[ii]
+                ii
                 for ii in itertools.chain.from_iterable(
                     smiles_indices[unique_smiles[i]] for i in val_idxs
                 )
             ]
             test = [
-                datapoints[ii]
+                ii
                 for ii in itertools.chain.from_iterable(
                     smiles_indices[unique_smiles[i]] for i in test_idxs
                 )
             ]
```
can't you just use `list()` here?

davidegraff commented on chemprop/v2/data/splitting.py:
```diff
@@ -206,9 +207,48 @@ def _unpack_astartes_result(
+    # split the data
+    train_idxs, val_idxs, test_idxs = split_data(datapoints, *args, **kwargs)
+
+    if isinstance(train_idxs[0], list):
+        # convert indices to datapoints for each fold
+        train = [[datapoints[i] for i in fold] for fold in train_idxs]
+        val = [[datapoints[i] for i in fold] for fold in val_idxs]
+        test = [[datapoints[i] for i in fold] for fold in test_idxs]
+    else:
+        # convert indices to datapoints
+        train = [datapoints[i] for i in train_idxs]
+        val = [datapoints[i] for i in val_idxs]
+        test = [datapoints[i] for i in test_idxs]
```
IMO type-checking should be avoided when possible:
1. it's non-pythonic: Python doesn't have strong typing support and goes against the core principle of duck typing
2. it obscures code meaning: I interpret this `if/else` block as branching depending on the result of `split_data`, which itself is a function of the arguments passed into the function. That is, depending on whether we're doing CV or not, we will get a result of `tuple[list[list[int]], ...]` or `tuple[list[int]], ...]`. We should just check that directly here by branching based on the value of `split_type`.

davidegraff reviewed (commented):
small comments

davidegraff commented on chemprop/v2/data/splitting.py:
```diff
@@ -31,7 +32,7 @@ def split_data(
     seed: int = 0,
     num_folds: int = 1,
-) -> tuple[Sequence[MoleculeDatapoint], ...]:
+) -> tuple[list[int], ...] | tuple[list[list[int], ...], ...]:
```
a return signature of `tuple[T, ...]` implies a tuple of unknown length containing items of type `T`. In the first return mode of this function, we should note that the tuple of is _known_ length (`== 3`) even if it's annoying to type out.

hwpang replied:
Thanks for the comment! It makes sense. I leave the type hints out as suggested in the comment above and move them into the docstring.

davidegraff commented on chemprop/v2/data/splitting.py:
```diff
@@ -206,9 +192,54 @@ def _unpack_astartes_result(
+            train = [[datapoints[i] for i in fold] for datapoints in datapointss for fold in train_idxs]
+            val = [[datapoints[i] for i in fold] for datapoints in datapointss for fold in val_idxs]
+            test = [[datapoints[i] for i in fold] for datapoints in datapointss for fold in test_idxs]
+        case SplitType.SCAFFOLD_BALANCED, SplitType.RANDOM_WITH_REPEATED_SMILES, SplitType.RANDOM, SplitType.KENNARD_STONE, SplitType.KMEANS:
```
is this formatted correctly?

hwpang replied:
Good catch! I missed a bracket in the list comprehension. I also notice that the syntax used in the case are not correct (should be `|` not `,`). I fixed them here and some other places.

davidegraff commented on chemprop/v2/data/splitting.py:
```diff
@@ -206,9 +192,54 @@ def _unpack_astartes_result(
+            val = [[datapoints[i] for i in val_idxs] for datapoints in datapointss]
+            test = [[datapoints[i] for i in test_idxs] for datapoints in datapointss]
+        case _:
+            raise ValueError(f'Split type "{split}" not supported.')
```
this should be a 
```suggestion
            raise RuntimeError("Unreachable code reached!")
```
the above call to `SplitType.get` should raise the appropriate `ValueError` and it should never fall through to this case, but it's always good practice to make sure that it doesn't.

hwpang replied:
Thanks! I committed this and also change the other similar exceptions.

davidegraff reviewed (commented):
I don't feel that strongly about the type hints even though I left a comment--I'm actually inclined to just leave them out because they make the source overly verbose. Thanks for taking care of this!

hwpang commented:
Thanks for the comments! I addressed all the comments and rebased with the newest `v2/dev` branch. I also fixed some minor bugs identified when I was testing out the cross validation for both monocomponent and multicomponent splitting.

davidegraff reviewed (approved):
👍  looks great! thanks for taking care of this
